pkg/storage/ingest/reader.go

```diff
@@ -0,0 +1,331 @@
+		if fetches.Err() != nil {
+			if errors.Is(fetches.Err(), context.Canceled) {
+				return nil
+			}
+			err := collectFetchErrs(fetches)
+			level.Error(r.logger).Log("msg", "encountered error while fetching", "err", err)
+			continue
+		}
+
+		r.recordFetchesMetrics(fetches)
+		r.consumeFetches(consumeCtx, fetches)
+		r.enqueueCommit(fetches)
+	}
+
+	return nil
+}
+
+func collectFetchErrs(fetches kgo.Fetches) (_ error) {
+	mErr := multierror.New()
+	fetches.EachError(func(s string, i int32, err error) {
+		// kgo advises to "restart" the kafka client if the returned error is a kerr.Error.
+		// Recreating the client would cause duplicate metrics registration, so we don't do it for now.
+		mErr.Add(err)
+	})
+	return mErr.Err()
+}
+
+func (r *PartitionReader) enqueueCommit(fetches kgo.Fetches) {
+	r.commitFetches <- fetches
+}
+
+func (r *PartitionReader) consumeFetches(ctx context.Context, fetches kgo.Fetches) {
+	records := make([]record, 0, len(fetches.Records()))
+
+	var minOffset, maxOffset int
+	fetches.EachRecord(func(record *kgo.Record) {
+		minOffset = min(minOffset, int(record.Offset))
+		maxOffset = max(maxOffset, int(record.Offset))
+		records = append(records, mapRecord(record))
+	})
+
+	err := r.consumer.consume(ctx, records)
+	if err != nil {
+		level.Error(r.logger).Log("msg", "encountered error processing records; skipping", "min_offset", minOffset, "max_offset", maxOffset, "err", err)
+		// TODO abort ingesting & back off if it's a server error, ignore error if it's a client error
+	}
+}
+
+func mapRecord(r *kgo.Record) record {
+	return record{
+		content:  r.Value,
+		tenantID: string(r.Key),
+	}
+}
+
+func (r *PartitionReader) recordFetchesMetrics(fetches kgo.Fetches) {
+	var (
+		now        = time.Now()
+		numRecords = 0
+	)
+
+	fetches.EachRecord(func(record *kgo.Record) {
+		numRecords++
+		r.metrics.receiveDelay.Observe(now.Sub(record.Timestamp).Seconds())
+	})
+
+	r.metrics.recordsPerFetch.Observe(float64(numRecords))
+}
+
+func (r *PartitionReader) newKafkaReader(at kgo.Offset) (*kgo.Client, error) {
+	client, err := kgo.NewClient(
+		kgo.ClientID(r.kafkaClientID),
+		kgo.SeedBrokers(r.kafkaAddress),
+		kgo.ConsumePartitions(map[string]map[int32]kgo.Offset{
+			r.kafkaTopic: {r.partitionID: at},
+		}),
+		kgo.FetchMinBytes(1),
+		kgo.FetchMaxBytes(100_000_000),
+		kgo.FetchMaxWait(5*time.Second),
+		kgo.FetchMaxPartitionBytes(50_000_000),
+		// Frequently refresh the cluster metadata. This allows us to quickly react in case some brokers are unhealthy,
+		// because the client updates the cluster info only at this frequency (and not after errors too).
+		//
+		// The side effect of frequently refreshing the cluster metadata is that if the backend returns each time a
+		// different authoritative owner for a partition, then each time cluster metadata is updated the Kafka client
+		// will client a new connection for each partition, leading to an high connections churn rate.
+		kgo.MetadataMaxAge(10*time.Second),
```

Comment:
Can you copy the commend and the setting (min age too) from the Writer, please? Or even better, defined a shared function returning common config options between the two clients, so that we don't copy it.